middleman_protocol/middleman_protocol/message.py

```diff
@@ -0,0 +1,196 @@
+import sys
+from hashlib import sha256
+
+from construct import Byte
+from construct import Bytes
+from construct import Const
+from construct import Container
+from construct import Enum
+from construct import GreedyBytes
+from construct import Int16ub
+from construct import Int32ub
+from construct import Prefixed
+from construct import Struct
+from construct import VarInt
+
+from golem_messages.message.base import Message as BaseGolemMessage
+
+from .constants import FRAME_SEPARATOR
+from .constants import PayloadType
+from .constants import PAYLOAD_TYPE_TO_MIDDLEMAN_MESSAGE_CLASS
+from .exceptions import MiddlemanProtocolError
+
+
+class AbstractMiddlemanMessage:
+
+    __slots__ = [
+        'payload',
+        'payload_type',
+        'request_id',
+    ]
+
+    def __init__(self, payload, request_id):
+        if type(self) == AbstractMiddlemanMessage:
+            raise NotImplementedError(
+                'AbstractMiddlemanMessage cannot be instantiated directly.'
+            )
+
+        self._validate_request_id(request_id)
+        self.request_id = request_id
+
+        self._validate_payload(payload)
+        self.payload = self._serialize_payload(payload)
+
+    @classmethod
+    def factory(cls, payload_type, payload, request_id):
+        assert payload_type in PayloadType
+        assert hasattr(sys.modules[__name__], PAYLOAD_TYPE_TO_MIDDLEMAN_MESSAGE_CLASS[payload_type])
+
+        return getattr(
+            sys.modules[__name__],
+            PAYLOAD_TYPE_TO_MIDDLEMAN_MESSAGE_CLASS[payload_type]
+        )(payload, request_id)
```

Comment:
This could use a comment. I think I know what and why you're trying to do but it's not obvious at a glance and also it may not be obvious why you're not just doing it in a different, possibly less magic way.

Reply:
Yes, adding comments is still listed in TODOs in this pull request so they will be added.